src/librustc/traits/auto_trait.rs

```diff
@@ -0,0 +1,646 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Support code for rustdoc and external tools . You really don't
+//! want to be using this unless you need to.
+
+use super::*;
+
+use std::collections::hash_map::Entry;
+use std::collections::VecDeque;
+
+use rustc_data_structures::fx::{FxHashMap, FxHashSet};
+
+use infer::region_constraints::{Constraint, RegionConstraintData};
+use infer::{InferCtxt, RegionObligation};
+
+use ty::fold::TypeFolder;
+use ty::{Region, RegionVid};
+
+// FIXME(twk): this is obviously not nice to duplicate like that
+#[derive(Eq, PartialEq, Hash, Copy, Clone, Debug)]
+pub enum RegionTarget<'tcx> {
+    Region(Region<'tcx>),
+    RegionVid(RegionVid),
+}
+
+#[derive(Default, Debug, Clone)]
+pub struct RegionDeps<'tcx> {
+    larger: FxHashSet<RegionTarget<'tcx>>,
+    smaller: FxHashSet<RegionTarget<'tcx>>,
+}
+
+pub enum AutoTraitResult<A> {
+    ExplicitImpl,
+    PositiveImpl(A),
+    NegativeImpl,
+}
+
+impl<A> AutoTraitResult<A> {
+    fn is_auto(&self) -> bool {
+        match *self {
+            AutoTraitResult::PositiveImpl(_) | AutoTraitResult::NegativeImpl => true,
+            _ => false,
+        }
+    }
+}
+
+pub struct AutoTraitInfo<'cx> {
+    pub full_user_env: ty::ParamEnv<'cx>,
+    pub region_data: RegionConstraintData<'cx>,
+    pub names_map: FxHashSet<String>,
+    pub vid_to_region: FxHashMap<ty::RegionVid, ty::Region<'cx>>,
+}
+
+pub struct AutoTraitFinder<'a, 'tcx: 'a> {
+    tcx: &'a TyCtxt<'a, 'tcx, 'tcx>,
+}
+
+impl<'a, 'tcx> AutoTraitFinder<'a, 'tcx> {
+    pub fn new(tcx: &'a TyCtxt<'a, 'tcx, 'tcx>) -> Self {
+        AutoTraitFinder { tcx }
+    }
+
+    /// Determine whether an auto trait is implemented for a type, and if this is the case if
+    /// non-trivial bounds need to be fulfilled, invoke a callback to compute a value representing
+    /// these in a fashion suitable for the caller.
```

Comment:
Can we update this comment a bit? I think I want it to say:

---

Make a best effort to determine whether and under which conditions an auto trait is implemented for a type. For example, if you have 

```rust
struct Foo<T> { data: Box<T> }
```

then this might return that `Foo<T>: Send` if `T: Send` (encoded in the `AutoTraitResult` type). The analysis attempts to account for custom impls as well as other complex cases. This result is intended for use by rustdoc and other such consumers.

(Note that due to the coinductive nature of `Send`, the full and correct result is actually quite simple to generate. That is, when a type has no custom impl, it is `Send` iff its field types are all `Send`. So, in our example, we might have that `Foo<T>: Send` if `Box<T>: Send`. But this is often not the best way to present to the user.)

Warning: The API should be considered highly unstable, and it may be refactored or removed in the future.